### PR TITLE
Security

### PR DESCRIPTION
**Pull request by Alexsandr-Yablunovskyi**
Added endpoints /register and /login. Assigning and roles process for user is implemented. Permission to endpoints by the users' role is implemented.

### CODE REVIEW

**Review by Cayman4010 — APPROVED** [review]
Cool solution!

File: src/main/java/yablunovskyi/bookstore/config/SecurityConfig.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+                .addFilterBefore(jwtAuthenticationFilter,
+                        UsernamePasswordAuthenticationFilter.class)
+                .userDetailsService(userDetailsService)
+                .build();
+    }
+    
+    @Bean
+    public AuthenticationManager authenticationManager(
+            AuthenticationConfiguration authenticationConfiguration
+    ) throws Exception {
```
Comment (NazarSmith228) [inline src/main/java/yablunovskyi/bookstore/config/SecurityConfig.java]:
```suggestion
            AuthenticationConfiguration authenticationConfiguration) throws Exception {
```

File: src/main/java/yablunovskyi/bookstore/controller/AuthenticationController.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import yablunovskyi.bookstore.dto.user.UserLoginResponseDto;
+import yablunovskyi.bookstore.dto.user.UserRegistrationRequestDto;
+import yablunovskyi.bookstore.dto.user.UserRegistrationResponseDto;
+import yablunovskyi.bookstore.exception.RegistrationException;
+import yablunovskyi.bookstore.security.AuthenticationService;
+import yablunovskyi.bookstore.service.UserService;
+
+@Tag(name = "Authentication process", description = "Endpoints to process user authentication")
+@RestController
+@Validated
```
Comment (NazarSmith228) [inline src/main/java/yablunovskyi/bookstore/controller/AuthenticationController.java]: you don't need this annotation here

File: src/main/java/yablunovskyi/bookstore/controller/AuthenticationController.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+@RequiredArgsConstructor
+@RequestMapping("/auth")
+public class AuthenticationController {
+    private final UserService userService;
+    private final AuthenticationService authenticationService;
+    
+    @PostMapping("/login")
+    public UserLoginResponseDto login(@RequestBody UserLoginRequestDto request) {
```
Comment (NazarSmith228) [inline src/main/java/yablunovskyi/bookstore/controller/AuthenticationController.java]: should you validate request body?

File: src/main/java/yablunovskyi/bookstore/controller/AuthenticationController.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        return authenticationService.authenticate(request);
+    }
+    
+    @PostMapping("/register")
+    public UserRegistrationResponseDto register(
+            @RequestBody @Valid UserRegistrationRequestDto request)
+            throws RegistrationException {
```
Comment (NazarSmith228) [inline src/main/java/yablunovskyi/bookstore/controller/AuthenticationController.java]:
```suggestion
            @RequestBody @Valid UserRegistrationRequestDto request) throws RegistrationException {
```

File: src/main/java/yablunovskyi/bookstore/controller/BookController.java [hunk 5]
```diff
@@ -33,6 +34,7 @@ public class BookController {
     @ResponseStatus(HttpStatus.OK)
     @Operation(summary = "Get all books",
             description = "Get a list of all available books from the datase")
+    @PreAuthorize("hasAnyRole('ADMIN', 'USER')")
```
Comment (NazarSmith228) [inline src/main/java/yablunovskyi/bookstore/controller/BookController.java:37]: shouldn't admin user also have a role `USER`?

or you design your app in a way that admin does not have same priviliges as user has?

File: src/main/java/yablunovskyi/bookstore/controller/BookController.java [hunk 6]
```diff
@@ -33,6 +34,7 @@ public class BookController {
     @ResponseStatus(HttpStatus.OK)
```
Comment (NazarSmith228) [inline src/main/java/yablunovskyi/bookstore/controller/BookController.java:34]: just FYI

`OK` is the default response status, so you can omit it

remember about this in future :)

File: src/main/java/yablunovskyi/bookstore/dto/user/UserLoginRequestDto.java [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+public record UserLoginRequestDto(
+        @NotEmpty
+        @Size(min = 8, max = 60)
+        @Email
+        String email,
+        
+        @NotEmpty
+                @Size(min = 8, max = 60)
+        String password
```
Comment (NazarSmith228) [inline src/main/java/yablunovskyi/bookstore/dto/user/UserLoginRequestDto.java]: the formatting here is really weird

pls fix it

File: src/main/java/yablunovskyi/bookstore/dto/user/UserRegistrationRequestDto.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+        @NotBlank(message = "can't be blank or null")
+        @Size(min = 8, max = 180, message = "length should be from 8 to 180")
+        String repeatPassword,
+        
+        @NotBlank(message = "can't be blank or null")
+        String firstName,
+        
+        @NotBlank(message = "can't be blank or null")
+        String lastName,
+        String shippingAddress
```
Comment (NazarSmith228) [inline src/main/java/yablunovskyi/bookstore/dto/user/UserRegistrationRequestDto.java:28]: shouldn't you validate this field?

File: src/main/java/yablunovskyi/bookstore/security/AuthenticationService.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+import yablunovskyi.bookstore.dto.user.UserLoginResponseDto;
+
+@Service
+@RequiredArgsConstructor
+public class AuthenticationService {
+    private final JwtUtil jwtUtil;
+    private final AuthenticationManager authenticationManager;
+    
+    public UserLoginResponseDto authenticate(UserLoginRequestDto request) {
+        final Authentication authentication = authenticationManager.authenticate(
```
Comment (NazarSmith228) [inline src/main/java/yablunovskyi/bookstore/security/AuthenticationService.java]:
```suggestion
        Authentication authentication = authenticationManager.authenticate(
```

File: src/main/java/yablunovskyi/bookstore/security/CustomUserDetailsService.java [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+
+@Component
+@RequiredArgsConstructor
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+    
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+        return userRepository.findByEmail(username).orElseThrow(
+                () -> new RuntimeException(
```
Comment (NazarSmith228) [inline src/main/java/yablunovskyi/bookstore/security/CustomUserDetailsService.java]: `RuntimeException` is wrong

File: src/main/java/yablunovskyi/bookstore/security/JwtAuthenticationFilter.java [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+            UserDetails userDetails = userDetailsService.loadUserByUsername(username);
+            Authentication authentication = new UsernamePasswordAuthenticationToken(
+                    userDetails, null, userDetails.getAuthorities());
+            SecurityContextHolder.getContext().setAuthentication(authentication);
+        }
+        filterChain.doFilter(request, response);
+    }
+    
+    private String getToken(HttpServletRequest request) {
+        String bearerToken = request.getHeader("Authorization");
```
Comment (NazarSmith228) [inline src/main/java/yablunovskyi/bookstore/security/JwtAuthenticationFilter.java]: there is an existing constant for authorization header

File: src/main/java/yablunovskyi/bookstore/service/impl/UserServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+    private final PasswordEncoder passwordEncoder;
+    private final RoleRepository roleRepository;
+    
+    @Override
+    public UserRegistrationResponseDto register(UserRegistrationRequestDto request)
+            throws RegistrationException {
+        if (userRepository.findByEmail(request.email()).isPresent()) {
```
Comment (NazarSmith228) [inline src/main/java/yablunovskyi/bookstore/service/impl/UserServiceImpl.java]: consider using `exists` syntax in the repository

**Review by NazarSmith228 — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution

File: src/main/java/yablunovskyi/bookstore/controller/BookController.java (hunk 6 above)
Comment (Alexsandr-Yablunovskyi) [reply src/main/java/yablunovskyi/bookstore/controller/BookController.java:34]: Nice, thanks!

File: src/main/java/yablunovskyi/bookstore/dto/user/UserRegistrationRequestDto.java (hunk 8 above)
Comment (Alexsandr-Yablunovskyi) [reply src/main/java/yablunovskyi/bookstore/dto/user/UserRegistrationRequestDto.java:28]: There was no requirement to a field "shippingAddress", so...

File: src/main/java/yablunovskyi/bookstore/controller/BookController.java (hunk 5 above)
Comment (Alexsandr-Yablunovskyi) [reply src/main/java/yablunovskyi/bookstore/controller/BookController.java:37]: Yes, ADMIN has the same priviliges as a USER, so to "findAll", "findById" methods I added @PreAuthorized("hasAnyRole('ADMIN', 'USER')") so ADMIN and USER can use this methods. And where only ADMIN has access to methods I add @PreAuthorize("hasRole('ADMIN')")

File: src/main/java/yablunovskyi/bookstore/security/JwtAuthenticationFilter.java (hunk 11 above)
Comment (Alexsandr-Yablunovskyi) [reply src/main/java/yablunovskyi/bookstore/security/JwtAuthenticationFilter.java]: Great, thanks!

File: src/main/java/yablunovskyi/bookstore/controller/AuthenticationController.java (hunk 2 above)
Comment (Elena-Bruyako) [reply src/main/java/yablunovskyi/bookstore/controller/AuthenticationController.java]: Not fixed

**Review by Elena-Bruyako — APPROVED** [review]
Good job! Fix one comment)

**Pull request merged**